torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -138,7 +138,108 @@ def get_tp_parallel_strategy(
     return RowwiseParallel, ColwiseParallel
 
 
-def apply_pipeline_parallelism(model, world_mesh, parallel_dims, job_config: JobConfig):
+class DummyTransformerLayer(torch.nn.Module):
+    def forward(self, input, freqs_cis):
+        return input
+
+
+class TransformerChunk(torch.nn.Module):
+    def __init__(
+        self,
+        orig_model,  # : Transformer,
+        this_stage_layer_names: List[str],
+        device,
+    ):
+        super().__init__()
+        self.tok_embeddings = None
+        if "tok_embeddings" in this_stage_layer_names:
+            self.tok_embeddings = orig_model.tok_embeddings
+
+        with torch.device(device):
+            self.freqs_cis = orig_model._precompute_freqs_cis()
+
+        # preserve FQNs of original model by preserving structure
+        # (including preserving position in layers[] list)- use dummy module
+        self.layers = orig_model.layers
+        for i in range(len(self.layers)):
+            if f"layers.{i}" not in this_stage_layer_names:
+                self.layers[i] = DummyTransformerLayer()
+        self.norm = None
+        if "norm" in this_stage_layer_names:
+            self.norm = orig_model.norm
+        self.output = None
+        if "output" in this_stage_layer_names:
+            self.output = orig_model.output
+
+    def forward(self, input):
+        """
+        Copypaste of original Transformer.forward, with conditionals and unpacking added
+        such that we handle the cases where this rank doesn't have the embedding, or doesn't have
+        the output layers.
+        """
+        if self.tok_embeddings:
+            h = self.tok_embeddings(input)
+        else:
+            h = input
+        _, seqlen, _ = h.shape
+
+        freqs_cis = self.freqs_cis[:seqlen]
+
+        for layer in self.layers:
+            h = layer(h, freqs_cis)
+        output = h
+
+        if self.norm:
+            h = self.norm(h)
+            output = h
+
+        if self.output:
```

Comment:
After looking into this, I feel these are some pretty common code, would love to brain storming this a bit more to see we could make this general enough

iiuc, this might be a bit tricky to generalize is that there's no easy way for us to know the output of a submodule would be look like, and whether it should just return the input directly without something like tracing?

Reply:
I have been thinking about this a lot too.

i think one proposal could be to introduce new building blocks that users can build their modules out of.  The two that come to mind are `OptionalLayer` and `OptionalModuleList`.

I haven't thought about this enough to write down a strong proposal though. I will try to do that.

Reply:
A downside to having 'OptionalLayer' is that it might "look" like the layer exists on the stage, when its actually disabled.  It'd be a shame if we registered a hook on the disabled layer, and the hook actually did something, which would be expected to only happen on some other stage where the layer actually exists.

The upside of it is that we can write simpler code in the top level model .forward() where we dont have to use conditionals.

At the end of the day, the transformer is simple enough that it might be OK to just check in the conditionals into the top level model, then hopefully I could totally delete the 'TransformerChunk' and 'DummyTransformerLayer' concepts.